src/subdomains/generic/gs/gs.service.ts

```diff
@@ -99,22 +100,47 @@ export class GsService {
     for (const userData of data) {
       const userDataId = userData.id ?? (userData['user_data_id'] as number);
 
+      const docsCache: { [b in string]: KycFile[] } = {};
+
       for (const selectPath of selectPaths) {
-        const prefixPath = this.toDocPrefixPath(selectPath, `${userDataId}`);
-
-        userData[selectPath] = prefixPath
-          ? await this.documentStorageService.listFilesByPrefix(prefixPath)
-          : [
-              ...(await this.documentStorageService.listUserFiles(userDataId)),
-              ...(await this.documentStorageService.listSpiderFiles(userDataId, false)),
-              ...(userData.accountType !== AccountType.PERSONAL
-                ? await this.documentStorageService.listSpiderFiles(userDataId, true)
-                : []),
-            ];
+        const docPath = this.toDocAffixPath(selectPath, `${userDataId}`);
+
+        const prefixPaths = this.getGCD(selectPath, selectPaths);
+        const docAffixPath = this.toDocAffixPath(prefixPaths, `${userDataId}`);
+
+        const docs =
+          docsCache[prefixPaths] ??
+          (docAffixPath
+            ? await this.documentStorageService.listFilesByPrefix(docAffixPath)
+            : [
+                ...(await this.documentStorageService.listUserFiles(userDataId)),
+                ...(await this.documentStorageService.listSpiderFiles(userDataId, false)),
+                ...(userData.accountType !== AccountType.PERSONAL
+                  ? await this.documentStorageService.listSpiderFiles(userDataId, true)
+                  : []),
+              ]);
+
+        if (!docsCache[prefixPaths]) docsCache[prefixPaths] = docs;
+
+        docPath === docAffixPath
+          ? (userData[selectPath] = docs)
+          : (userData[selectPath] = docs.filter((doc) => doc.url.includes(docPath)));
       }
     }
   }
 
+  private getGCD(inputString: string, selects: string[], separator = '.'): string {
+    const gcd = [];
+
+    // or with for loop?
+    inputString.split(separator).forEach((separatedInput, i) => {
+      const t = selects.every((s) => s.split(separator)[i] === separatedInput);
+      t && i == gcd.length && gcd.push(separatedInput);
+    });
+
+    return gcd.length === 0 ? 'documents' : gcd.join(separator);
+  }
```

Comment:
Use prefix on character basis, see e.g. https://stackoverflow.com/questions/68702774/longest-common-prefix-in-javascript